src/Nethermind/Nethermind.Merge.Plugin/Handlers/ForkchoiceUpdatedHandler.cs

```diff
@@ -464,4 +439,52 @@ private BlockInfo[] GetBeaconChainBranch(Block newHeadBlock, BlockInfo newHeadBl
 
         return blocksList.ToArray();
     }
+
+    private class StateRootUpdater : IDisposable
+    {
+        private readonly IWorldState _globalWorldState;
+        private readonly IBlockProcessingQueue _queue;
+        private Hash256? _stateRoot;
+
+        public StateRootUpdater(IWorldState globalWorldState, IBlockProcessingQueue queue)
+        {
+            _globalWorldState = globalWorldState;
+            _queue = queue;
+            _queue.ProcessingQueueEmpty += OnProcessingQueueEmpty;
+        }
+
+        public Hash256? StateRoot
+        {
+            set
+            {
+                if (_stateRoot != value)
+                {
+                    Interlocked.Exchange(ref _stateRoot, value);
+                    if (value is not null && _queue.IsEmpty)
+                    {
+                        UpdateStateRoot();
```

Comment:
@LukaszRozmej would you consider removing this fast path for the update? The reason would be to always run on the queue empty handler rather than potentially make an in-flight cross-thread change. It should not change much in performance (we finally have the queue empty and we run on it) but could address some potential issues.

Reply:
Tehn you can potentially wait next 12s

Reply:
But we progress. The lack of moving forward was the initial cause to implement the update of the checkpoint. I don't see anything wrong with waiting for 1 block if this makes the threading reasoning simpler. Maybe I'm missing something here.

Reply:
well there is logic in what you say. I did want to make it consistent with FCU's received though.

Reply:
BTW if queue is empty then the lock on NewPayload/FCU should prevent anything to be added to the queue